File: Data Structures & Algorithms/generate-parentheses/submission-0.java
class Solution {
    public List<String> generateParenthesis(int n) {
        List<String> res = new ArrayList<>();
        Stack<Character> goodParen = new Stack<>();
        backTrack(0, 0, n, goodParen, res);
        return res;
    }
    private void backTrack(int openC, int closedC, int totalPairs, Stack<Character> stack, List<String> res){
        StringBuilder sB = new StringBuilder();
        if(openC == closedC && closedC + openC == (totalPairs*2)){
            for(char c : stack){
                sB.append(c);
            }
            res.add(sB.toString());
            return;
        }
        if(openC < totalPairs){
            stack.push('(');
            backTrack(openC+1, closedC, totalPairs, stack, res);
            stack.pop();
        }
        if(closedC < openC){
            stack.push(')');
            backTrack(openC, closedC+1, totalPairs, stack, res);
            stack.pop();
        }
    }
}

File: Data Structures & Algorithms/subsets/submission-0.py
class Solution:
    def subsets(self, nums: List[int]) -> List[List[int]]:
        res = []

        def backtrack(index, curr_set):
            res.append(curr_set[:])
            for i in range(index, len(nums)):
                curr_set.append(nums[i])
                backtrack(i + 1, curr_set)
                curr_set.pop()
        
        backtrack(0, [])
        return res
        